rfcs/0004-categorized-attributes.md

```diff
@@ -0,0 +1,217 @@
+reverse thrust
+turn
+active cooling
+ramscoop
+cloak
+cooling
+force
+energy
+fuel
+heat
+jam
+```
+
+It MAY include any or all of the following:
+```
+disabled
+minable
+piercing
+```
+
+The supported _modifiers_ SHOULD include all of the following:
+```
+none
+multiplier
+relative
+over time
+```
+
+The _default modifier_ MUST be `none`.
+
+### Implementation
+
+The constraints of attribute values MUST be enforced in runtime. _Categorized attributes_ with their default value MAY NOT be stored. There MUST exist an abstraction layer supporting queries and updates of stored and non-stored _categorized attributes_. This abstraction layer SHOULD support non-_categorized attributes_ as well. This layer MUST support storing any _category_+_effect_+_modifier_ combination as a _categorized attribute_, regardless of engine support for such an attribute.
+
+Querying the value of a non-stored attribute MUST return its default value, and MUST NOT cause exceptions or undefined behaviour. 
+
+An implementation MUST allow efficient query of all _categorized attributes_ when given a _category_. It MUST allow for efficient transformation of all _categorized attributes_ of a given category, such as adding another set of matching attributes to it on a per-value basis.
+
+For instance, such an operation MUST be performed efficiently:
+
+```
+capacity
+	shield 100
+	hull 20
+
+-
+
+damage
+	shield 10
+	hull 10
+
+=
+
+capacity
+	shield 90
+	hull 10
+```
+
+An implementation SHOULD store _categorized attributes_ grouped by _category_. It SHOULD NOT store the _category_ of the individual attribute values.
+
+An implementation MAY use a similar structure for storing attributes:
+
+```
+class Entry {
+	Effect e;
+	Modifier m;
+}
+
+std::map<Entry, double> attributes[CATEGORY_COUNT];
+```
+```
+class Entry {
+	Effect e;
+	Modifier m;
+
+	double value;
+}
+
+std::map<Category, std::set<Entry>> attributes;
+```
+
+An implementation MUST provide const iterator functionality over attributes. It MAY provide non-const iterator functionality.
+
+An implementation MUST provide efficient functionality matching that of Outfit::CanAdd for _categorized attributes_.
+
+# Drawbacks
+
+## Maintenance cost:
+
+We have to keep a list of all _old-form_ attribute names that the engine refers to using the new format. Though this list need not be updated or maintained, it represents an additional cost.
+
+This change will also heavily conflict with and other pull requests that create new _categorized attributes_, or change engine code using such attributes. They will all have to be moved to the _new-form_ syntax, creating a burden on the PR authors.
+
+## Performance
+
+Though the performance of the final form of this RFC should be better than our current performance, it would perform significantly worse while only the data format is implemented. Early tests suggest a doubling of the individual lookup cost of categorized attributes using the currently implemented method in the worst case.
```

Comment:
I disagree that performance would be worse when only the new data format (keywords) are implemented; we can always use the new format with the old implementation (and convert at load-time from the new format back to the old format).